### ISSUE TITLE
Prevent overwriting existing test(s) by mistakenly using existing test tag for new Cucumber scenarios

### ISSUE THREAD

**Issue by atddkg** [Qytera-Gmbh/cypress-xray-plugin#335]
Hi,

My intention is to use our automation frameworks/IDE as the single source of truth for test ticket creation which I've discovered poses a possible problem: the plugin requires that I tag my Cucumber scenarios (e.g. `@TestName:PRJ-123`) but I'm going to want to make these tags unique so as not to overwrite existing tests with new tests... seemingly placing the burden on automation engineers manually checking one or more frameworks and/or Jira to ensure they're not using an existing test tag.

Does the plugin cater for this problem?

Thanks

**Comment by csvtuda** [comment]
Hi @atddkg,

thanks for the issue. Let's see if I understood your problem correctly.

---

First, let's assume you initially have a Cucumber scenario in a feature file which is linked to a Jira issue:

```cucumber
Feature: Abc

    @TestName:CYP-123
    Scenario: Xyz
        Given ...
        When ...
        Then ...
```

Next, you start working on a new test, duplicate the existing scenario (copy the file) but mistakenly forget to update the scenario's tag:


```cucumber
Feature: New

    @TestName:CYP-123
    Scenario: Very new
        Given ...
        When ...
        Then ...
```

The plugin will then overwrite `CYP-123`, which is obviously not what you want. Is this correct?

---

If yes, then there is currently nothing to prevent that. However, it would actually be possible to implement safeguards against this. One such safeguard could be to simply display a warning and do nothing with `CYP-123`. Would that work for you?

If this is not what you meant, please describe your problem in more detail.

**Comment by atddkg** [comment]
I think what you're talking about is in the scope of a framework - detecting duplicate tags within a specific framework, is that correct?

I'm thinking of checking if any of the tags within a framework exist on tickets in Jira and to warn if so. Ideally, I'd also like to have the option of abandoning test ticket creation/update & test run if test tag(s) exist on ticket(s) in Jira, informing the automation engineer to resolve the detected duplicates. But this could be an option to switch on/off because there might be a use case I haven't thought of whereby automation engineers need to use the same test tag on multiple scenarios.

Perhaps a better way to deal with this is, for scenarios that don't have a test tag, have Xray/Jira create a test ticket and return the ticket ID & scenario title to the automation engineer so they can manually or otherwise map them after the fact (or better, automatically decorate/annotate the scenarios with the ticket ID as a test tag) - thereby putting the burden of test tag uniqueness on Jira, where it is more strongly enforced.

**Comment by csvtuda** [comment]
I think this topic deserves a longer answer from me.

> I'm thinking of checking if any of the tags within a framework exist on tickets in Jira and to warn if so.

> I think what you're talking about is in the scope of a framework - detecting duplicate tags within a specific framework, is that correct?

Not quite, I'm talking about detecting duplicate tags within a single execution of Cypress.

The plugin _only_ runs when a Cypress execution is triggered and only uses the following data (currently):
- the feature files to run in the execution (through the `on("file:preprocessor")` hook)
  - used to import the scenarios contained within to Xray, _after_ checking that they contain proper tags to prevent new issues from being created
- the test results such as passes and failures, screenshots and videos

It doesn't scan the project when you save, modify or create files and so on and thus cannot warn you about existing Jira issues before you even run Cypress. This is something you'd need to solve outside of the plugin's scope, e.g. a plugin for VS Code (if this is what you wish for - not sure).

> I'd also like to have the option of abandoning test ticket creation/update & test run if test tag(s) exist on ticket(s) in Jira

The plugin is unfortunately designed in exactly the opposite way. It prevents file/scenario uploads if the files do _not_ contain existing issue keys 😕

I know that the initial effort to get the plugin running is quite high (_"I need to manually tag every feature file/create new Jira issues if Cypress was there first? Bummer!"_), but having it create issues on the fly is very complicated territory (see below or [the referenced issues in this response](https://github.com/Qytera-Gmbh/cypress-xray-plugin/issues/336#issuecomment-2153146933)).

> Perhaps a better way to deal with this is, for scenarios that don't have a test tag, have Xray/Jira create a test ticket and return the ticket ID & scenario title to the automation engineer so they can manually or otherwise map them after the fact (or better, automatically decorate/annotate the scenarios with the ticket ID as a test tag) - thereby putting the burden of test tag uniqueness on Jira, where it is more strongly enforced.

On paper, it sounds easy enough. But it would open the box of pandora for sure:
- What if the provided credentials do not have permissions to create issues? 
- What if there are mandatory fields which need to be configured when creating new Jira issues (environment, parent user story, assignee, status, ...)?
  - The plugin would need to expose [configuration](https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-issues/#api-rest-api-3-issue-post) to the user on how Jira issues should be created. In other words, again placing the burden on the test automation engineer (_"Okay, if the scenario title contains `Login`, then the test needs to be assigned to Bob, ..."_)
- What if there's an autoformatter configured for feature files, or other stuff such as tabs vs. spaces?

**TL;DR:** Please keep in mind that the plugin is more or less supposed to simply upload test results. It's not meant to be a fully-fletched Cypress/Jira/Xray integration with feature file scanning (or even more advanced features such as tapping into Xray's Gherkin step library for step verification). I'd like to keep it simple (also because I'm just one guy with a little bit of time to spare here and there for developing it).

---

On the bright side, the plugin already displays verbose warnings for the following cases:

- a feature file does not contain issue key tags: https://github.com/Qytera-Gmbh/cypress-xray-plugin/blob/9d7b78b6af3891ce488fdd87d60d5613e489b454/src/preprocessing/preprocessing.spec.ts#L235-L266

- a feature file was tagged properly and uploaded, but the Xray response contained different issue keys: https://github.com/Qytera-Gmbh/cypress-xray-plugin/blob/9d7b78b6af3891ce488fdd87d60d5613e489b454/src/hooks/preprocessor/synchronizeFeatureFile.spec.ts#L205-L222

Unfortunately, duplicate tag detection does not exist yet. But it could be done for sure and display warnings for feature files that contain the same tags on distinct backgrounds/scenarios. 
This at least I could implement for you.

**Comment by atddkg** [comment]
After experimentation, the concerns I raised in this thread have been relatively unfounded. Jira seems pretty robust identifying the correct Test ticket to update despite using the same ID in the scenario tags (note: I've moved to using raw IDs for tags, e.g. @TST-n (without a prefix)).
